File: benchmarks/64-image-resize/image_resize.m
(* 64-image-resize -- ImageResize against skimage.

   Downsampling uses AREA AVERAGING on both sides (Mathilda's Automatic, skimage's
   downscale_local_mean), and upsampling uses bilinear with anti-aliasing off, so the two are
   doing the same arithmetic rather than approximately similar arithmetic.

   As in 63-image-convolve the image is built outside the timed region: constructing 262144 Expr
   reals is not resampling, and timing it would say so. *)

n = 512;
img = Image[Table[N[Mod[x*7 + y*13, 251]/251], {y, n}, {x, n}]];

bench["IR area 512->256", ImageResize[img, {256, 256}];];
bench["IR area 512->128", ImageResize[img, {128, 128}];];
bench["IR nearest 512->256", ImageResize[img, {256, 256}, Resampling -> "Nearest"];];
bench["IR bilinear 512->1024", ImageResize[img, {1024, 1024}];];

(* Area averaging PRESERVES THE MEAN exactly at an integer factor, because every source pixel
   contributes with equal total weight. A conservation law, so it is a check rather than a
   timing -- and it is what fails if the coverage weights are ever normalised wrongly. *)
check["IR area preserves mean",
  Round[10^9 (Mean[Flatten[ImageData[ImageResize[img, {256, 256}]]]]
              - Mean[Flatten[ImageData[img]]])]];

(* THE aliasing discriminator. A 4x4 checkerboard to 2x2: area averaging gives 0.5 everywhere,
   nearest gives a flat field with the pattern annihilated. *)
check["IR checkerboard area total",
  Round[10^6 Total[Flatten[ImageData[ImageResize[
    Image[{{0.,1.,0.,1.},{1.,0.,1.,0.},{0.,1.,0.,1.},{1.,0.,1.,0.}}], {2, 2}]]]]];

check["IR block mean exact",
  Round[10^9 Total[Flatten[ImageData[ImageResize[
    Image[{{0.,0.25,0.5,0.75},{0.,0.25,0.5,0.75},{1.,1.,1.,1.},{1.,1.,1.,1.}}], {2, 2}]]]]];

File: benchmarks/64-image-resize/image_resize.py
"""64-image-resize -- skimage as the baseline for ImageResize.

downscale_local_mean is the right comparison for the area path: it is exactly a block mean at an
integer factor, which is what Mathilda's area averaging reduces to when the fractional coverage
terms vanish. skimage's `resize` with anti_aliasing=True would prefilter with a Gaussian first and
so measure a different algorithm.

For the enlargement, resize(order=1, anti_aliasing=False) is plain bilinear, matching Mathilda's
default for enlargement (there are no frequencies to remove when enlarging).
"""
import numpy as np
from skimage.transform import resize, downscale_local_mean


def ramp(n):
    y, x = np.mgrid[1:n + 1, 1:n + 1]
    return ((x * 7 + y * 13) % 251) / 251.0


img = ramp(512)

bench("IR area 512->256", lambda: downscale_local_mean(img, (2, 2)))
bench("IR area 512->128", lambda: downscale_local_mean(img, (4, 4)))
bench("IR nearest 512->256", lambda: resize(img, (256, 256), order=0, anti_aliasing=False))
bench("IR bilinear 512->1024", lambda: resize(img, (1024, 1024), order=1, anti_aliasing=False))

small = downscale_local_mean(img, (2, 2))
check("IR area preserves mean", round(1e9 * (small.mean() - img.mean())))

chk = np.array([[0., 1., 0., 1.], [1., 0., 1., 0.],
                [0., 1., 0., 1.], [1., 0., 1., 0.]])
check("IR checkerboard area total", round(1e6 * downscale_local_mean(chk, (2, 2)).sum()))

q = np.array([[0., .25, .5, .75], [0., .25, .5, .75],
              [1., 1., 1., 1.], [1., 1., 1., 1.]])
check("IR block mean exact", round(1e9 * downscale_local_mean(q, (2, 2)).sum()))
